### PR TITLE
Rename workspaces check if out name exists

### LINKED ISSUE #16263: RenameWorkspaces should only overwrite if forced
### Expected behaviour

RenameWorkspace should fail if an existing workspace of the output name is defined.  This check needs to be for anything in the ADS and also members of groups in the ADS.

We should add a new overwite Boolean property (default false) that would delete the outputworkspace if it exists before renaming.

### Actual behaviour
Currently renameWorkspace can overwite workspaces if the outputworkspace name is already in use.
This can make a particular mess if you name two members within a workspace group the same (see #8652).

### Steps to reproduce the behavior
```python 
CreateSampleWorkspace(OutputWorkspace='bob', Function='User Defined', UserDefinedFunction='name=LinearBackground,A0=3.22542,A1=-0.0633193')
CreateSampleWorkspace(OutputWorkspace='bob1', Function='User Defined', UserDefinedFunction='name=LinearBackground,A0=3.22542,A1=-0.0633193')
CreateSampleWorkspace(OutputWorkspace='bob2', Function='User Defined', UserDefinedFunction='name=LinearBackground,A0=3.22542,A1=-0.0633193')
CreateSampleWorkspace(OutputWorkspace='bob3', Function='User Defined', UserDefinedFunction='name=LinearBackground,A0=3.22542,A1=-0.0633193')
GroupWorkspaces(InputWorkspaces='bob,bob1,bob2,bob3', OutputWorkspace='NewGroup')
RenameWorkspace(InputWorkspace='bob1', OutputWorkspace='bob')
```
This will give you a group with two members with the same name, there is lots of trouble from this point on.

### Platforms affected
All

Once this is solved we should check with Rob Dalgleish to make sure this does not mess up his scripts.

### PR DESCRIPTION
**Pull request by DavidFair**
Description of work.
Modified RenameWorkspace and RenameWorkspaces to check if any of the output names are used in other workspaces
**To test:**
Renaming a workspace to the same name as an existing workspace should throw an error
Renaming workspaces should throw an error if any of the output names match an existing name

Fixes #16263 .

Release notes have been updated

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [x] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/RenameWorkspaces.cpp [hunk 1]
```diff
@@ -86,6 +96,18 @@ void RenameWorkspaces::exec() {
     }
   }
 
+  // Check all names are not used already before starting rename
+  for (size_t i = 0; i < nWs; ++i) {
```
Comment (FedeMPouzols) [inline Framework/Algorithms/src/RenameWorkspaces.cpp]: I think here you could use a range-based loop: `for (const auto &name : newWsName) { ...`

File: Framework/Algorithms/src/RenameWorkspace.cpp [hunk 2]
```diff
@@ -44,13 +52,28 @@ void RenameWorkspace::exec() {
...
   // get the output workspace name
   std::string outputwsName = getPropertyValue("OutputWorkspace");
+  // check if we are overriding existing workspaces
+  bool overrideWorkspaces = getProperty("OverrideExisting");
 
   if (getPropertyValue("InputWorkspace") ==
       getPropertyValue("OutputWorkspace")) {
     throw std::invalid_argument(
         "The input and output workspace names must be different");
   }
```
Comment (FedeMPouzols) [inline Framework/Algorithms/src/RenameWorkspace.cpp]: This check could be done at the `validateInputs()` step.

File: Framework/Algorithms/src/RenameWorkspace.cpp [hunk 3]
```diff
@@ -44,13 +52,28 @@ void RenameWorkspace::exec() {
...
+    // Output name already exists - either remove or error
+    if (!overrideWorkspaces) {
+      // If we try to delete the workspace here a subtle bug is introduced
+      // Where the workspace group handle is deleted if we are renaming
+      // Its last workspace member, then when we add (or rename) that member
+      // undefined behavior happens usually Python Unit tests breaking
+      throw std::runtime_error("The workspace " + outputwsName +
+                               " already exists");
+    }
+  }
```
Comment (FedeMPouzols) [inline Framework/Algorithms/src/RenameWorkspace.cpp]: And probably this too can go in `validateInputs()`?

**Comment by FedeMPouzols** [comment]
@DavidFair: I've made some comments for possible improvements,. let me know what you think. But I think this looks good and ready to go. It doesn't break backwards compatibility, code looks good, and the documentation and examples are very good.

**Comment by DavidFair** [comment]
I'll move the examples into validateInputs .
Previously the code also deleted the target workspace at execution time to make sure however we later discovered this caused a race condition where Poco would delete the workspace handle which we then returned so now we let the ADS handle it.

**Comment by DavidFair** [comment]
@FedeMPouzols I've moved all tests (except one) into validate inputs and reworked the unit tests to use these now.
In RenameWorkspaces the check to see if names are in use has to be done at execution time as we need to join suffix + name + prefix then query the ADS for each one.

File: docs/source/release/v3.8.0/framework.rst [hunk 4]
```diff
@@ -49,6 +49,10 @@ Improved
 - :ref:`SetSample <algm-SetSample>`: Fixed a bug with interpreting the `Center` attribute for cylinders/annuli 
 
 
+- :ref:`RenameWorkspace <algm-RenameWorkspace>` and `RenameWorkspaces <algm-RenameWorkspaces>`
+  now check if a Workspace with that name already exists in the ADS and gives
+  the option to override it.
```
Comment (FedeMPouzols) [inline docs/source/release/v3.8.0/framework.rst]: override => overwrite? If you modify this we don't need to wait for the builds again.

**Comment by FedeMPouzols** [comment]
Looks good. It implements the plan in the issue description, code looks good, it is well covered with new/extended unit tests, and the documentation reads well.
Keeping a default of this new option as "yes, overwrite" makes perfect sense for backwards compatibility.

:shipit:

**Comment by martyngigg** [comment]
@DavidFair This branch has developed a merge conflict

**Pull request merged**